rescript/replace_taxonomy.py

```diff
@@ -0,0 +1,50 @@
+# ----------------------------------------------------------------------------
+# Copyright (c) 2021, QIIME 2 development team.
+#
+# Distributed under the terms of the Modified BSD License.
+#
+# The full license is in the file LICENSE, distributed with this software.
+# ----------------------------------------------------------------------------
+
+from qiime2.plugin import MetadataColumn, Categorical
+import pandas as pd
+import re
+
+
+def make_substitutions_dict(taxonomy_replacement_map):
+    replacement_dict = taxonomy_replacement_map.to_series().to_dict()
+    return replacement_dict
+
+
+def make_regex(substitutions):
+    strings = substitutions.keys()
+    regex = re.compile('|'.join(map(re.escape, strings)))
+    return regex
+
+
+def lineages_with_invalid_number_of_ranks(updated_tax_series,
+                                          num_expected_ranks):
+    invalid_lineages = [label for label, lineage in updated_tax_series.items()
+                        if len(lineage.split(';')) != num_expected_ranks]
+    return invalid_lineages
+
+
+def replace_taxonomy(taxonomy: pd.Series,
+                     taxonomy_replacement_map: MetadataColumn[Categorical],
+                     num_expected_ranks: int = 7,
+                     ) -> pd.Series:
+
+    rd = make_substitutions_dict(taxonomy_replacement_map)
+    mr = make_regex(rd)
+
+    updated_tax_series = taxonomy.apply(lambda taxonomy:
```

Comment:
why use re.sub for this at all? Wouldn't [pd.Series.replace](https://pandas.pydata.org/docs/reference/api/pandas.Series.replace.html) work? It uses `re.sub` under the hood so might save some effort (and potentially make the code more efficient), and looks like it can take a regex, dict, series, or other input to specify the substitutions. Then the preceding lines/functions could be dropped entirely, and just do something like:
```
updated_tax_series = taxonomy.replace(taxonomy_replacement_map)
```

Reply:
'cause I'm a dummy 😄  . I just had to add `regex=True` n order for it to work on substrings. Otherwise it'll try and match the whole string. I got this working, and added a flag to toggle `regex`.

I removed all the other code too.

Reply:
oh good find... maybe this option should be exposed so that users can define whether to match entire or partial strings.

Is there no way to to substring substitution while `regex=False`? I worry a little bit about users unintentionally using wildcards that should be interpreted literally, e.g., brackets or punctuation in some taxonomies...

maybe your approach is more sound after all if it gives more flexibility...

Reply:
>Is there no way to to substring substitution while regex=False? 

I think we'd have to use `str.replace` or something along those lines. Then use a `lambda` function...

Reply:
>I worry a little bit about users unintentionally using wildcards that should be interpreted literally, e.g., brackets or punctuation in some taxonomies...

Yeah, me too, which is why I was using `re.escape`. But not sure how to implement that with the `pd.Series.replace`. I was initially thinking of having a flag that would toggle `re.escape` and `re`.

Reply:
I figured out a way to pass find-replace patterns to `pd.replace()` in the way I was imagining. We can now toggle whether or not the find pattern passes through `re.escape` prior to being sent to `re.compile`.  That resulting dictionary of compiled regexes can then be fed to `pd.replace()`. 🎉